core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,139 @@
+          "provided by the underlying file format.");
+    }
+    Preconditions.checkNotNull(dataEncryptionConfig,
+        "Cannot create EnvelopeEncryptionManager because data encryption config is not passed");
+    this.dataEncryptionConfig = dataEncryptionConfig;
+    Preconditions.checkNotNull(dataEncryptionConfig.kekId(),
+        "Cannot create EnvelopeEncryptionManager because table key encryption key ID is not specified");
+    Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.kmsClient = kmsClient;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    if (nativeFormatEncryption) {
+      NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
+          .encryptionAlgorithm(metadata.algorithm())
+          .build();
+
+      Preconditions.checkArgument(rawOutput instanceof NativelyEncryptedFile,
+          "Can't natively encrypt " + rawOutput.location() + " because the class " +
+              rawOutput.getClass() + " doesn't implement NativelyEncryptedFile interface");
+
+      ((NativelyEncryptedFile) rawOutput).setNativeCryptoParameters(nativeEncryptParams);
+
+      return new BaseEncryptedOutputFile(rawOutput, metadata);
+    } else {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only file formats which " +
+          "have native encryption implementations");
+    }
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    EnvelopeMetadata metadata = EnvelopeMetadataParser.fromJson(encrypted.keyMetadata().buffer());
+    ByteBuffer fileDek = kmsClient.unwrapKey(metadata.wrappedDek(), metadata.kekId());
+
+    if (nativeFormatEncryption) {
+      // Pushdown to data formats with native decryption support (such as Parquet)
+      NativeFileCryptoParameters nativeDecryptParams = NativeFileCryptoParameters.create(fileDek).build();
+      InputFile rawInput = encrypted.encryptedInputFile();
+
+      Preconditions.checkArgument(rawInput instanceof NativelyEncryptedFile,
+          "Can't natively decrypt " + rawInput.location() + " because the class " +
+              rawInput.getClass() + " doesn't implement NativelyEncryptedFile interface");
+
+      ((NativelyEncryptedFile) rawInput).setNativeCryptoParameters(nativeDecryptParams);
+
+      return rawInput;
+    } else {
+      throw new UnsupportedOperationException("EnvelopeEncryptionManager currently supports only file formats which " +
+          "have native encryption implementations");
+    }
+  }
+
+  private EnvelopeMetadata generateEnvelopeMetadata(EnvelopeConfiguration config) {
+    String tableKekID = config.kekId();
+    byte[] fileDek;
+    String wrappedFileDEK;
+
+    if (kmsClient.supportsKeyGeneration()) {
```

Comment:
Should this call move to initialize time and save it somewhere to use here? A given KMS won't dynamically changed whether or not it can support KeyGeneration, but if we have RPC calls every time we call encrypt() seems is a waste.

Reply:
sgtm